Compiling with GHC 8.0.1-rc4

The below patch allows compiling with GHC 8.0.1-rc4. Sadly it does so via the use of UndecidableInstances. I did not see a quick way to avoid its use since:

class HasDynamicDimension a => HasDimension a where
instance (KnownDimension d) => HasDynamicDimension (Dimensional v d a) where
type KnownDimension (d :: Dimension) = HasDimension (Proxy d)

leads to the clear loop (expanded):

class HasDynamicDimension a => HasDimension a where
instance (HasDimension (Proxy d)) => HasDynamicDimension (Dimensional v d a) where

This patch avoids https://ghc.haskell.org/trac/ghc/ticket/12026

```
diff --git a/src/Numeric/Units/Dimensional/Dimensions/TypeLevel.hs b/src/Numeric/Units/Dimensional/Dimensions/TypeLevel.hs
index 51e07b4..c52493d 100644
--- a/src/Numeric/Units/Dimensional/Dimensions/TypeLevel.hs
+++ b/src/Numeric/Units/Dimensional/Dimensions/TypeLevel.hs
@@ -38,7 +38,7 @@ where

 import Data.Proxy
 import Numeric.NumType.DK.Integers
-  ( TypeInt (Zero, Pos1, Pos2, Pos3), (+)(), (-)()
+  ( TypeInt (Zero, Pos1, Pos2, Pos3), type (+)(), type (-)()
   , KnownTypeInt, toNum
   )
 import qualified Numeric.NumType.DK.Integers as N
diff --git a/src/Numeric/Units/Dimensional/Dynamic.hs b/src/Numeric/Units/Dimensional/Dynamic.hs
index b48c192..dde9f37 100644
--- a/src/Numeric/Units/Dimensional/Dynamic.hs
+++ b/src/Numeric/Units/Dimensional/Dynamic.hs
@@ -17,6 +17,7 @@ Defines types for manipulation of units and quantities without phantom types for
 {-# LANGUAGE KindSignatures #-}
 {-# LANGUAGE NoImplicitPrelude #-}
 {-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE UndecidableInstances #-}

 module Numeric.Units.Dimensional.Dynamic
 (
@@ -83,7 +84,7 @@ instance (KnownDimension d) => Demotable (Quantity d) where
       -- This implementation is not provided directly inside the instance because it requires ScopedTypeVariables
       -- Placing the signatures inside the instance requires InstanceSigs, which interacts poorly with associated type families
       -- like Dimensional in GHC 7.8.
-      demoteQ :: forall a (d' :: Dimension).KnownDimension d' => AnyQuantity a -> Maybe (Quantity d' a)
+      demoteQ :: forall a (d' :: Dimension). AnyQuantity a -> Maybe (Quantity d' a)
       demoteQ (AnyQuantity dim val) | dim == dim' = Just . Quantity $ val
                                     | otherwise   = Nothing
         where
diff --git a/src/Numeric/Units/Dimensional/Internal.hs b/src/Numeric/Units/Dimensional/Internal.hs
index 6feac7b..0f91717 100644
--- a/src/Numeric/Units/Dimensional/Internal.hs
+++ b/src/Numeric/Units/Dimensional/Internal.hs
@@ -207,11 +207,11 @@ instance (KnownDimension d, Show a, Fractional a) => Show (Quantity d a) where
 --
 -- >>> showIn watt $ (37 *~ volt) * (4 *~ ampere)
 -- "148.0 W"
-showIn :: (KnownDimension d, Show a, Fractional a) => Unit m d a -> Quantity d a -> String
+showIn :: (Show a, Fractional a) => Unit m d a -> Quantity d a -> String
 showIn (Unit n _ y) (Quantity x) | Name.weaken n == nOne = show (x / y)
                                  | otherwise             = (show (x / y)) ++ " " ++ (show n)

-instance (KnownDimension d, Show a) => Show (Unit m d a) where
+instance (Show a) => Show (Unit m d a) where
   show (Unit n e x) = "The unit " ++ show n ++ ", with value " ++ show e ++ " (or " ++ show x ++ ")"

 -- Operates on a dimensional value using a unary operation on values, possibly yielding a Unit.
```

The changes in TypeLevel.hs are required because GHC got more picky about namespaces, and types are different namespace. It must always be specified now. The code previously crashed GHC, as per the bug, but this does not.

UndecidableInstances is required, to avoid:

```
src/Numeric/Units/Dimensional/Dynamic.hs:79:10: error:
    • The constraint ‘KnownDimension d’
        is no smaller than the instance head
      (Use UndecidableInstances to permit this)
    • In the instance declaration for ‘Demotable (Quantity d)’
```

The type signature changes are just warning cleanups since the instances were not used. They are completely unrelated to either bug.

Does that help?

This error you mentioned above:

```
src/Numeric/Units/Dimensional/Dynamic.hs:79:10: error:
    • The constraint ‘KnownDimension d’
        is no smaller than the instance head
      (Use UndecidableInstances to permit this)
    • In the instance declaration for ‘Demotable (Quantity d)’
```

seems inaccurate? The constraint `KnownDimension d` expands to `HasDimension (Proxy d)`. How is that no smaller than the instance head `Quantity d` = `Dimensional 'DQuantity d`? I guess I don't know what the language lawyer definition of "smaller" here is, but it seems OK in the sense that we can make progress without guessing, because we erased the bits about `Dimensional` and `'DQuantity` and now we just need to go off and investigate `HasDimension (Proxy d)`, which will lead us to https://github.com/bjornbm/dimensional/blob/master/src/Numeric/Units/Dimensional/Dimensions/TypeLevel.hs#L144-L151, and this appears to be how 7.8.4 and 7.10.3 proceed.

I'm digging through the GHC change list to try to figure out what bug in 7.10 we were exploiting to get away with this that they have shut down.

Thanks again for taking the time to report this and dig in to it! I hope dimensional has been helpful to your work in some way.

--- a/.gitignore
+++ b/.gitignore
@@ -1,5 +1,6 @@
 dist
 .cabal-sandbox/
+.stack-work/
 cabal.sandbox.config
 *.hi
 *.o

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,8 @@
+1.0.1.2 (2016-05)
+-----------------
+* Support for GHC 8.0.1-rc4, avoiding GHC Trac issue 12026.
+* Added support for stack.
+
 1.0.1.1 (2015-11)
 -----------------
 * Improved example in readme.

--- a/dimensional.cabal
+++ b/dimensional.cabal
@@ -1,5 +1,5 @@
 name:                dimensional
-version:             1.0.1.1
+version:             1.0.1.2
 license:             BSD3
 license-file:        LICENSE
 copyright:           Bjorn Buckwalter 2006-2015

--- a/src/Numeric/Units/Dimensional/Dimensions/TypeLevel.hs
+++ b/src/Numeric/Units/Dimensional/Dimensions/TypeLevel.hs
@@ -38,7 +38,7 @@ where
 
 import Data.Proxy
 import Numeric.NumType.DK.Integers
-  ( TypeInt (Zero, Pos1), (+)(), (-)()
+  ( TypeInt (Zero, Pos1), type (+), type (-)
   , KnownTypeInt, toNum
   )
 import qualified Numeric.NumType.DK.Integers as N

--- a/stack.yaml
+++ b/stack.yaml
@@ -1,0 +1,14 @@
+# For more information, see: https://github.com/commercialhaskell/stack/blob/release/doc/yaml_configuration.md
+
+# Specifies the GHC version and set of packages available (e.g., lts-3.5, nightly-2015-09-21, ghc-7.10.2)
+resolver: lts-4.1
+
+# Local packages, usually specified by relative directory name
+packages:
+- '.'
+
+# Packages to be pulled from upstream that are not in the resolver (e.g., acme-missiles-0.3)
+extra-deps: []
+
+# Override default flag values for local packages and extra-deps
+flags: {}

